840 short description for each video

karjo24 opened the pull request:
### Motivation and Context
#840. Having the ability to make short notes or add/change the title of the lecture is useful for organising lectures when titles are not provided by the course admins.

### Description
There's an additional option in the dropdown menu for each stream on the course overview page that allows users to input their own (limited to 80 characters) lecture title. There's a new API endpoint, that takes the new title and stores it in a new table in the database.

### Steps for Testing
Prerequisites:
- 1 User
- 1 Course with 1 stream with title, 1 stream without

1. Log in and navigate to course
2. Click on the "More options" menu (three dots next to stream)
3. Click on "Edit personal lecture title"
4. Enter new title in popup menu
5. Verify the following behaviour
  5.1 Pressing "Enter" or clicking outside the input field changes the title
  5.2 Pressing "Escape" disregards any changes
  5.3 Deleting the value of the input field resets the title to the actual title provided by the lecturer
6. Reload to make sure the changes are persistent

### Screenshots

https://github.com/user-attachments/assets/d8adfd72-19e3-43b7-ab91-e7ec0bc55373

karjo24 committed:
button for personal lecture title change

karjo24 committed:
frontend structure

karjo24 committed:
removed readonly

karjo24 committed:
database base setup of new table

karjo24 committed:
cleaned up dao, added updateorcreate

karjo24 committed:
extended backend to include custom lecture naem

karjo24 committed:
query fix

karjo24 committed:
frontend includes customName, only api missing

karjo24 committed:
renamed function

karjo24 committed:
added api

karjo24 committed:
added api call

karjo24 committed:
fixed wrong cache setting

karjo24 committed:
fixed eslint suggestions

karjo24 committed:
adapted tests to changes of this PR; Eslint fix

karjo24 committed:
golang lint fix

karjo24 committed:
golang linter fix

karjo24 committed:
deletes personal lecture name when set to empty string

carlobortolan commented on api/courses.go:
```diff
@@ -344,7 +344,14 @@ func (r coursesRoutes) getCourseBySlug(c *gin.Context) {
 	}
 
-	course, err := r.CoursesDao.GetCourseBySlugYearAndTerm(c, uri.Slug, query.Term, query.Year)
+	user := tumLiveContext.User
+	var course model.Course
+	var err error
+	if user == nil {
+		course, err = r.CoursesDao.GetCourseBySlugYearAndTerm(c, uri.Slug, query.Term, query.Year)
+	} else {
+		course, err = r.CoursesDao.GetCourseBySlugYearTermUser(c, uri.Slug, query.Term, query.Year, user.ID)
+	}
```
Maybe consider handling both cases (with and without user) in the GetCourseBySlugYearAndTerm method to avoid duplicate code in `GetCourseBySlugYearAndTerm` and `GetCourseBySlugYearTermUser` (see comment below)

karjo24 replied:
Fixed nearly everything, thanks for the review :)

carlobortolan commented on dao/courses.go:
```diff
@@ -231,6 +232,25 @@ func (d coursesDao) GetCourseBySlugYearAndTerm(ctx context.Context, slug string,
 }
 
+// GetCourseBySlugYearTermUser loads courses with streams preloaded. Difference to GetCourseBySlugYearAndTerm: personal lecture titles of user are loaded as well
+func (d coursesDao) GetCourseBySlugYearTermUser(ctx context.Context, slug string, term string, year int, userID uint) (model.Course, error) {
```
Would it be possible to simplify both `GetCourseBySlug...` methods into one or would it break something?

E.g., something like:
```go
func (d coursesDao) GetCourseBySlugYearAndTerm(slug string, term string, year int, userID uint) (model.Course, error) {
    var cacheKey string
    if userID == 0 {
        cacheKey = fmt.Sprintf("courseBySlugYearAndTerm%v%v%v", slug, term, year)
    } else {
        cacheKey = fmt.Sprintf("courseBySlugYearTermUser%v%v%v%v", slug, term, year, userID)
    }

    cachedCourses, found := Cache.Get(cacheKey)
    if found {
        return cachedCourses.(model.Course), nil
    }

    var course model.Course
    query := DB.Preload("Streams.VideoSections").Preload("Streams.Units", func(db *gorm.DB) *gorm.DB {
        return db.Order("unit_start desc")
    }).Preload("Streams", func(db *gorm.DB) *gorm.DB {
        return db.Order("start desc")
    }).Preload("Admins").Where("teaching_term = ? AND slug = ? AND year = ?", term, slug, year)

    if userID != 0 {
        query = query.Preload("Streams.CustomLectureTitles", "user_id = ?", userID)
    }

    err := query.First(&course).Error
    if err == nil {
        Cache.SetWithTTL(cacheKey, course, 1, time.Minute)
    }
    return course, err
}
```

karjo24 replied:
Yes, that probably makes sense. As this only adds information to the return struct, I doubt that this breaks something (I did have a look at all usages and did a quick test locally). Technically we'd not need the check for userID != 0, because the where clause would not match any entry for == 0, but I'd like to keep it for clarity.

carlobortolan commented on web/template/home.gohtml:
```diff
@@ -511,6 +517,12 @@
+                                                                                    @click="showLectureTitleInput=true;vod.Dropdown.value=false;"
+                                                                                    class="tum-live-menu-item">
+                                                                                <i class="fa-solid fa-pencil-square mr-4"></i>
+                                                                                <span class="block">Edit personal lecture title</span>
```
Consider renaming this to "Personalize title" or "Edit personal title" to make the text shorter and similar to the other video actions.
![image](https://github.com/user-attachments/assets/6c50b837-7241-453e-b198-477bfac55d17)

carlobortolan reviewed (commented):
Tested it locally and it looks good to me :+1: 

Just two more general questions: 
- Should the personal lecture title be displayed only in the course-view, or also in the "Recent VODs" and videoplayer/stream-view?
- Should lecturers/admins of a course be allowed to add personal lecture titles, or would it be better to have the button set the actual lecture title to avoid confusion?

carlobortolan committed:
Add custom lecture title to apiv2.proto

karjo24 committed:
shorter menu entry title for conciseness

karjo24 committed:
refactored two getCourseBySlugYearAndTerm[User] into one

karjo24 committed:
changed tests to match new function signature

karjo24 commented:
>     * Should the personal lecture title be displayed only in the course-view, or also in the "Recent VODs" and videoplayer/stream-view?

Yes, thanks for catching that. Somehow thought the title was not displayed there and didn't check. I'll change that.
 
>     * Should lecturers/admins of a course be allowed to add personal lecture titles, or would it be better to have the button set the actual lecture title to avoid confusion?

It probably shouldn't be left like this as I agree with the confusion part. I'm not sure, whether we should hide the menu option for admins or make it change the title globally. After all, that's what the admin page is for. I lean towards the "hide it"-approach, but I'm open for other opinions.

carlobortolan commented:
> It probably shouldn't be left like this as I agree with the confusion part. I'm not sure, whether we should hide the menu option for admins or make it change the title globally. After all, that's what the admin page is for. I lean towards the "hide it"-approach, but I'm open for other opinions.

I think we can leave it out for now and add it in the future in a separate PR if necessary.

joschahenningsen commented on api/courses.go:
```diff
@@ -344,7 +344,14 @@ func (r coursesRoutes) getCourseBySlug(c *gin.Context) {
 	}
 
-	course, err := r.CoursesDao.GetCourseBySlugYearAndTerm(c, uri.Slug, query.Term, query.Year)
+	userId := uint(0)
```
nit:
```suggestion
	var userId uint
```

joschahenningsen commented on dao/user_defined_lecture_titles.go:
```diff
@@ -0,0 +1,50 @@
+package dao
```
Nit: I know we do it in some places, but go discurages this style of filename. My suggestion would be `lecture_titles.go`

joschahenningsen commented on dao/user_defined_lecture_titles.go:
```diff
@@ -0,0 +1,50 @@
+	return d.db.Delete(&model.UserDefinedLectureTitle{}, "user_id = ? AND stream_id = ?", userID, streamID).Error
+}
+
+// Upsert updates the entry if it exists, inserts it else
+func (d userDefinedLectureTitlesDao) Upsert(userLectureTitle *model.UserDefinedLectureTitle) error {
+	return d.db.Save(userLectureTitle).Error
+}
```
Do we need upserting? With user_id and stream_id being composite primary keys, we can use create with a conflict clause to update a record. See e.g. https://github.com/TUM-Dev/gocast/blob/fe0190d8b3d1c6e536bc4e5cb8f82cc2e5a534b2/dao/progress.go#L58
I would go that way and rename the method to `Save` and drop `Create` and `Upsert`

karjo24 replied:
Done.

joschahenningsen commented on model/stream.go:
```diff
@@ -398,9 +400,14 @@ func (s Stream) ToDTO() StreamDTO {
 		duration = s.Duration.Int32
 	}
+	customLectureName := ""
+	if len(s.CustomLectureTitles) > 0 {
+		customLectureName = s.CustomLectureTitles[0].Title
+	}
```
What exactly is happening here? We just use the first title? Wouldn't we send a random users title to someone?

karjo24 replied:
Although this isn't set to a random users title, this should definitely be changed, yes, Idk what I was thinking. It wasn't set to a random title, because the CustomLectureTitles in this PR only get preloaded in conjunction with a Where clause, that checks the userId.

karjo24 replied:
It is now fixed. However, is there a cleaner way using Gorm to achieve the same to avoid having to add it manually afterwards?

joschahenningsen commented on model/user_defined_lecture_titles.go:
```diff
@@ -0,0 +1,23 @@
+// BeforeCreate is a GORM hook that is called before a new user is created.
+// UserDefinedLectureTitle will not be saved if the title is too long
+func (u *UserDefinedLectureTitle) BeforeCreate(tx *gorm.DB) (err error) {
+	if len(u.Title) > 256 {
```
Can you trim the whitespace? https://pkg.go.dev/strings#TrimSpace

karjo24 replied:
Is there any particular reason for this? Wouldn't this lead to a title being falsely accepted by `BeforeCreate` and then failing `Save`, if #whitespace + #meaningfulLetters > 256, but #meaningfulLetters < 256? Is this a best practice I'm not aware of?

joschahenningsen commented on dao/courses.go:
```diff
@@ -214,19 +214,34 @@ func (d coursesDao) GetInvitedUsersForCourse(course *model.Course) error {
+func (d coursesDao) GetCourseBySlugYearAndTerm(ctx context.Context, slug string, term string, year int, userID uint) (model.Course, error) {
+	var cacheKey string
+	if userID == 0 {
+		cacheKey = fmt.Sprintf("courseBySlugYearAndTerm%v%v%v", slug, term, year)
+	} else {
+		cacheKey = fmt.Sprintf("courseBySlugYearTermUser%v%v%v%v", slug, term, year, userID)
```
Not going to be an issue soon if we don't have weird years, but this introduces a cache key collision if e.g. year=1, and userId=11 we produce the same key as year=11, userId=1. I suggest adding a `-` in between

joschahenningsen reviewed (commented):
Pretty nice! Left just a few comments

karjo24 committed:
code style changes

karjo24 committed:
refactored upsert to save using clauses

karjo24 committed:
changed file name and edited error message to reflect prior changes

karjo24 committed:
changed where lecture title gets added to stream DTO to avoid future bugs

karjo24 committed:
changed cache key to avoid collisions

karjo24 committed:
Merge branch 'dev' into 840-short-decription-for-each-video

SebiWrn commented:
Maybe there should also be a button for the user to delete the custom title, as just removing it to reset it to the lecturers title may not be intuitive.

karjo24 committed:
incorrect redirect removed

karjo24 committed:
eslint fix

Copilot committed:
Trim whitespace from custom lecture title before saving

Agent-Logs-Url: https://github.com/TUM-Dev/gocast/sessions/3e52aecf-8cbc-4bde-a8dc-d086283d8439

Co-authored-by: alexanderstephan <30267166+alexanderstephan@users.noreply.github.com>

joschahenningsen commented:
Sadly closing this as stale
